ssl: Add an option to disable peer host verification

orgads opened the pull request.

bagder commented:
Please provide a description for your pull request as well, not only in the commit.

How is this different than `CURLOPT_SSL_VERIFYHOST`?

and for development systems, you should mostly avoid disabling checks and instead change the name/IP with `CURLOPT_RESOLVE` or similar.

orgads commented:
I can't find a command-line flag that enables only `CURLOPT_SSL_VERIFYHOST`. We'll try `CURLOPT_RESOLVE`, thanks for the reference. Closing for now.

orgads commented:
I only rebased it and pushed as a draft to see if the checks pass. I'll need to ask the maintainer why this is needed. Will try to do it next week.

orgads commented:
> Is there a reason why `-k`/`--insecure` doesn't solve this specific bad SSL case?

It does, but it's too permissive. It will succeed even if you have a self-signed certificate, while this new option won't.

vszakats commented:
Given that `-k`/`--insecure` is a big enough issue as it is already. If anything, a new variant would better beging with `--insecure-*` to make its nature obvious. But, I'm not sure the world needs one more option to disable TLS security features.

orgads commented:
Ok, I now reverted all the changes in the lib, and renamed the flag to `--ssl-no-verify-host`. All it does it unsetting `CURLOPT_SSL_VERIFYHOST` without unsetting `CURLOPT_SSL_VERIFYPEER` (`--insecure` unsets both).

Notice that it is based on #13704 to avoid merge conflicts. If the other change is not accepted, I can rebase and remove this dependency.

Is this acceptable?

vszakats commented:
I remain convinced that curl doesn't need another insecure option. Also the new option name makes it even more difficult to detect if a script/app is using any insecure option, and also doesn't make it clear that this option breaks TLS security, hence my suggestion to begin with the `--insecure` prefix.

orgads commented:
> I remain convinced that curl doesn't need another insecure option. Also the new option name makes it even more difficult to detect if a script/app is using any insecure option, and also doesn't make it clear that this option breaks TLS security, hence my suggestion to begin with the `--insecure` prefix.

So `--insecure-ssl-no-verify-host`?

orgads commented:
By the way, `--ssl-no-revoke` is also "less secure", but doesn't have an explicit prefix.

vszakats commented:
> > I remain convinced that curl doesn't need another insecure option. Also the new option name makes it even more difficult to detect if a script/app is using any insecure option, and also doesn't make it clear that this option breaks TLS security, hence my suggestion to begin with the `--insecure` prefix.
> 
> So `--insecure-ssl-no-verify-host`?

Or, `--insecure-host`?

bagder commented:
I still lack a proper motivation for this new option. `--insecure`, `--resolve` and `--connect-to` should be able to fulfill most use cases. And if they don't, I want to learn exactly why.

vszakats commented:
> By the way, `--ssl-no-revoke` is also "less secure", but doesn't have an explicit prefix.

That's true. Luckily this has a more limited impact being Schannel-specific anf revocation generally being outphased in favour of short expiration.

orgads commented:
> I still lack a proper motivation for this new option. `--insecure`, `--resolve` and `--connect-to` should be able to fulfill most use cases. And if they don't, I want to learn exactly why.

I tried `--resolve` and `--connect-to`, but couldn't get them to accept https://wrong.host.badssl.com. Can you suggest a way?

`--insecure` is too permissive, as I wrote earlier. I do want to check that the certificate is signed properly, but not verify the host name. I'm still checking with the maintainer if they have a real-life scenario this covers. Will update.

bagder commented:
> --insecure is too permissive, as I wrote earlier. I do want to check that the certificate is signed properly, but not verify the host name. I'm still checking with the maintainer if they have a real-life scenario this covers. Will update.

Without a name check, verifying the certificate is close to pointless. There are several hundred million HTTPS sites in the world and certificates for all of them can be digitally verified. If you don't check the name, your server can just use any of those certificates without you knowing. It really is very close to as insecure as `--insecure`. So no, I don't think adding such an option helps any user. It does not make your connection safe but will keep it at roughly the `--insecure` level. So you can just as well use `--insecure`.

> I tried --resolve and --connect-to, but couldn't get them to accept https://wrong.host.badssl.com/. Can you suggest a way?

I don't understand what you want to do.

orgads commented:
We're still trying to figure out what was the intention behind this (it was done in 2011 😅), and if we can come up with a better solution for our use-case. I'm closing it for now.

orgads commented:
Reopened just to rebase. You may ignore this.

orgads commented:
Reopening for rebase again.

bagder commented:
This PR was deemed not for us. If you want to rebase or otherwise work on it, please do that in your own fork.
